app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -230,38 +228,53 @@ func processMetrics(metrices <-chan []byte, contentType expfmt.Format) []byte {
 		}
 	}
 
-	str := processNewlineChars(buf.String(), true, true)
-	str = fmt.Sprintf("%s\n", str)
+	processedMetrics := processNewlineChars(buf.Bytes(), true, true)
+	processedMetrics = append(processedMetrics, '\n')
+	buf.Reset()
+	_, err := buf.Write(processedMetrics)
+	if err != nil {
+		panic(err)
+	}
 
-	// if the content type is not OpenMetrics, we don't need to add EOF marker
-	if !(contentType == expfmt.FmtOpenMetrics_1_0_0 || contentType == expfmt.FmtOpenMetrics_0_0_1) {
-		return []byte(str)
+	if contentType == expfmt.FmtOpenMetrics_1_0_0 || contentType == expfmt.FmtOpenMetrics_0_0_1 {
+		// make metrics OpenMetrics compliant
+		buf.Write([]byte("# EOF\n"))
+		return buf.Bytes()
 	}
 
-	// make metrics OpenMetrics compliant
-	return []byte(fmt.Sprintf("%s# EOF\n", str))
+	return buf.Bytes()
 }
 
-// processNewlineChars processes the newline characters in the text.
-// If dedup is true, it replaces multiple newline characters with a single newline character.
+// processNewlineChars processes the newline characters in the byteData.
+// If dedup is true, it replaces multiple consecutive newline characters with a single newline character.
 // If trim is true, it trims the leading and trailing newline characters.
-func processNewlineChars(text string, dedup, trim bool) string {
+func processNewlineChars(byteData []byte, dedup, trim bool) []byte {
```

Comment:
let's remove these parameters since we don't actually use them

Reply:
Okay.. So in that case, shall we consider having one liner `bytes.TrimSpace(dedup(byteData, '\n', '\n'))` in `processMetrics()` and removing `processNewlineChars()` altogether and its unit-tests?

Reply:
Amm, I am just keeping `processNewlineChars()` even if what is does is just a one liner to test combination of `bytes.TrimSpace()` and `dedup()`